### PR TITLE
Validate _error task before submission

### PR DESCRIPTION
**Pull request by frsyuki**
`_error` tasks are generated when a task fail. Unlike other dynamically
generated tasks, generating `_error` tasks can't notice its failure by
making the task failed because the task is already failed.

This change adds catch sections to WorkflowExecutor around
`compiler.compileTasks`. Because _error task is now validated, the new
catch for `_error` tasks is not necessary, strictly speaking. However,
here takes more defensive approach because throwing an uncaught
exception in workflow executor loop means that workflow execution
stops.

### CODE REVIEW

**Comment by komamitsu** [comment]
@frsyuki 
It seems the following workflow doesn't work with this change.
```
+task:
    fail>: Failed.....!
    _error:
        echo>: Something is wrong with something
```
Is this an expected behaviour?
```
[mitsu@onion mydig]$ ~/src/digdag/pkg/digdag-0.9.14-SNAPSHOT.jar run mydig.dig
2017-08-02 14:36:52 +0900: Digdag v0.9.14-SNAPSHOT
error: io.digdag.core.repository.ModelValidationException: Validating workflow task failed
name must not contain character '+': "++mydig+task^error" (config)
```
0.9.13 and `master` branch handle the workflow successfully.
```
[mitsu@onion mydig]$ digdag run mydig.dig
2017-08-02 14:36:39 +0900: Digdag v0.9.13
2017-08-02 14:36:41 +0900 [WARN] (main): Using a new session time 2017-08-02T00:00:00+00:00.
2017-08-02 14:36:41 +0900 [INFO] (main): Using session /Users/mitsu/tmp/mydig/.digdag/status/20170802T000000+0000.
2017-08-02 14:36:41 +0900 [INFO] (main): Starting a new session project id=1 workflow name=mydig session_time=2017-08-02T00:00:00+00:00
2017-08-02 14:36:42 +0900 [INFO] (0017@[0:default]+mydig+task): fail>: Failed.....!
2017-08-02 14:36:42 +0900 [ERROR] (0017@[0:default]+mydig+task): Task +mydig+task failed.
Failed.....!
2017-08-02 14:36:43 +0900 [INFO] (0017@[0:default]+mydig+task^error): echo>: Something is wrong with something
Something is wrong with something
2017-08-02 14:36:43 +0900 [INFO] (0017@[0:default]+mydig^failure-alert): type: notify
error:
  * +mydig+task:
    Failed.....!

Task state is saved at /Users/mitsu/tmp/mydig/.digdag/status/20170802T000000+0000 directory.
  * Use --session <daily | hourly | "yyyy-MM-dd[ HH:mm:ss]"> to not reuse the last session time.
  * Use --rerun, --start +NAME, or --goal +NAME argument to rerun skipped tasks.
```

File: digdag-tests/src/test/resources/acceptance/error/invalid.dig [hunk 1]
```diff
@@ -0,0 +1,7 @@
++message:
+  echo>: Workflow started
++task:
+  echo>: ok
+  _error:
+    echo>: invalid
+    sh>: echo duplicated
```
Comment (komamitsu) [inline digdag-tests/src/test/resources/acceptance/error/invalid.dig]: Considering https://github.com/treasure-data/digdag/pull/610#issuecomment-319573687, maybe we'd better add another test case that uses a nested `_error` with proper configuration and see if it finishes successfully.

**Comment by frsyuki** [comment]
@komamitsu that's bad. I added another commit and fixed the issue with a new test.

**Comment by komamitsu** [comment]
@frsyuki This change doesn't detect a configuration error of the following workflow in compile phase?

```
$ cat mydig.dig
+task:
    fail>: Failed.....!

_error:
    echo>: Something is wrong with something
    sh>: echo Something is wrong with something
```

```
$ ~/src/digdag/pkg/digdag-0.9.14-SNAPSHOT.jar run mydig
2017-08-03 13:51:40 +0900: Digdag v0.9.14-SNAPSHOT
2017-08-03 13:51:41 +0900 [WARN] (main): Using a new session time 2017-08-03T00:00:00+00:00.
2017-08-03 13:51:41 +0900 [INFO] (main): Using session /Users/mitsu/tmp/mydig/.digdag/status/20170803T000000+0000.
2017-08-03 13:51:41 +0900 [INFO] (main): Starting a new session project id=1 workflow name=mydig session_time=2017-08-03T00:00:00+00:00
2017-08-03 13:51:42 +0900 [INFO] (0017@[0:default]+mydig+task): fail>: Failed.....!
2017-08-03 13:51:42 +0900 [ERROR] (0017@[0:default]+mydig+task): Task +mydig+task failed.
Failed.....!
2017-08-03 13:51:42 +0900 [WARN] (main): Found a broken _error task in attempt 1 task 1. Skipping this task.
io.digdag.client.config.ConfigException: A task can't have more than one operator: {"echo>":"Something is wrong with something","sh>":"echo Something is wrong with something","_export":{"error":{"message":"Failed.....!"}}}
        at io.digdag.core.workflow.WorkflowCompiler$Context.collect(WorkflowCompiler.java:181)
        at io.digdag.core.workflow.WorkflowCompiler$Context.compile(WorkflowCompiler.java:131)
        at io.digdag.core.workflow.WorkflowCompiler.compileTasks(WorkflowCompiler.java:51)
        at io.digdag.core.workflow.WorkflowExecutor.addErrorTasksIfAny(WorkflowExecutor.java:1393)
        at io.digdag.core.workflow.WorkflowExecutor.setDoneFromDoneChildren(WorkflowExecutor.java:647)
        at io.digdag.core.workflow.WorkflowExecutor.lambda$null$16(WorkflowExecutor.java:575)
        at io.digdag.core.database.DatabaseSessionStoreManager.lambda$lockTaskIfExists$19(DatabaseSessionStoreManager.java:433)
        at io.digdag.core.database.BasicDatabaseStoreManager.transaction(BasicDatabaseStoreManager.java:163)
        at io.digdag.core.database.DatabaseSessionStoreManager.lockTaskIfExists(DatabaseSessionStoreManager.java:427)
        at io.digdag.core.workflow.WorkflowExecutor.lambda$null$17(WorkflowExecutor.java:574)
        at io.digdag.core.database.ThreadLocalTransactionManager.begin(ThreadLocalTransactionManager.java:250)
        at io.digdag.core.database.ThreadLocalTransactionManager.begin(ThreadLocalTransactionManager.java:220)
        at io.digdag.core.workflow.WorkflowExecutor.lambda$propagateAllPlannedToDone$18(WorkflowExecutor.java:573)
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:474)
        at io.digdag.core.workflow.WorkflowExecutor.propagateAllPlannedToDone(WorkflowExecutor.java:576)
        at io.digdag.core.workflow.WorkflowExecutor.runWhile(WorkflowExecutor.java:510)
        at io.digdag.core.workflow.WorkflowExecutor.runUntilDone(WorkflowExecutor.java:456)
        at io.digdag.core.LocalSite.runUntilDone(LocalSite.java:162)
        at io.digdag.cli.Run.run(Run.java:293)
        at io.digdag.cli.Run.run(Run.java:242)
        at io.digdag.cli.Run.main(Run.java:173)
        at io.digdag.cli.Main.cli(Main.java:187)
        at io.digdag.cli.Main.main(Main.java:82)
2017-08-03 13:51:42 +0900 [INFO] (0017@[0:default]+mydig^failure-alert): type: notify
error:
  * +mydig+task:
    Failed.....!
```

**Comment by komamitsu** [comment]
Ah, that's an expected one. > `Found a broken _error task in attempt 1 task 1. Skipping this task.`

**Comment by komamitsu** [comment]
LGTM 👍  although I think there is a room for improvement.

1. When an `_error` task that has configuration errors exists as a subtask, digdag shows an error message w/o any stack trace. On the other hand,  when such a `_error` task exists as a child of a root task, digdag shows stack trace. I think It would reduce confusion if the difference is resolved.

2. It'd be great if you add a test that confirm the digdag execution loop doesn't stop even if `_error` task fails.

**Comment by frsyuki** [comment]
oh really...? It doesn't happen to me...
```
$ cat x.dig
+task:
  fail>: Failed.....!

_error:
  echo>: Something is wrong with something
  sh>: echo Something is wrong with something

$ ./pkg/digdag-0.9.14-SNAPSHOT.jar run x.dig
2017-08-03 11:22:30 -0700: Digdag v0.9.14-SNAPSHOT
error: A task can't have more than one operator: {"echo>":"invalid","sh>":"echo duplicated"} (config)
```

**Comment by frsyuki** [comment]
I see! "digdag run" command reads and compiles all *.dig files on the current directory.

**Comment by frsyuki** [comment]
Validation was missing at grouping task. Added the validation and test case.

**Comment by frsyuki** [comment]
`Found a broken _error task in attempt 1 task 1. Skipping this task.` error message shouldn't happen usually. It happens only either of under following scenario:

* Persistent DB (PostgreSQL or H2) is used. Old version of digdag stored invalid workflow before without validation. New version executed it.
* There is an unexpected bug.

**Pull request merged**